File: web/lib/api/workforce.ts
import { API_PREFIX, apiFetchApp } from "./backend-client";

async function parseError(response: Response): Promise<string> {
  try {
    const payload = (await response.json()) as { detail?: string };
    return payload.detail ?? `Request failed with status ${response.status}`;
  } catch {
    return `Request failed with status ${response.status}`;
  }
}

export type EmployeeSummary = {
  id: string;
  business_id: string;
  primary_location_id?: string | null;
  primary_location_name?: string | null;
  primary_role_id?: string | null;
  primary_role_name?: string | null;
  external_ref?: string | null;
  employee_number?: string | null;
  full_name: string;
  preferred_name?: string | null;
  phone_e164?: string | null;
  email?: string | null;
  base_hourly_rate_cents?: number | null;
  date_of_birth?: string | null;
  minor_school_status?: string | null;
  work_permit_number?: string | null;
  work_permit_effective_start_on?: string | null;
  work_permit_expires_on?: string | null;
  work_permit_max_daily_minutes?: number | null;
  work_permit_max_weekly_minutes?: number | null;
  work_permit_earliest_start_local_time?: string | null;
  work_permit_latest_end_local_time?: string | null;
  reliability_score?: number | null;
  status: string;
  employment_type?: string | null;
  hire_date?: string | null;
  termination_date?: string | null;
  notes?: string | null;
  employee_metadata: Record<string, unknown>;
  notification_preferences?: EmployeeNotificationPreferences;
  role_ids: string[];
  role_names: string[];
  location_ids: string[];
  location_names: string[];
  created_at: string;
  updated_at: string;
};

export type EmployeeNotificationPreferences = {
  schedule_publish_email_enabled: boolean;
  schedule_publish_sms_enabled: boolean;
  email_opted_out_at?: string | null;
  sms_opted_out_at?: string | null;
  email_opt_out_reason?: string | null;
  sms_opt_out_reason?: string | null;
};

export type EmployeeWorkPermitWeekday =
  | "sunday"
  | "monday"
  | "tuesday"
  | "wednesday"
  | "thursday"
  | "friday"
  | "saturday";

export type EmployeeWorkPermitRuleProfile = {
  template_code?: string | null;
  jurisdiction_code?: string | null;
  source_url?: string | null;
  allowed_weekdays?: EmployeeWorkPermitWeekday[];
  daily_max_minutes?: number | null;
  daily_max_minutes_in_session?: number | null;
  daily_max_minutes_summer_break?: number | null;
  daily_max_minutes_school_day?: number | null;
  daily_max_minutes_non_school_day?: number | null;
  daily_max_minutes_preceding_non_school_day?: number | null;
  weekly_max_minutes?: number | null;
  weekly_max_minutes_in_session?: number | null;
  weekly_max_minutes_summer_break?: number | null;
  weekly_max_minutes_school_week?: number | null;
  weekly_max_minutes_non_school_week?: number | null;
  earliest_start_local_time?: string | null;
  earliest_start_local_time_school_day?: string | null;
  earliest_start_local_time_non_school_day?: string | null;
  latest_end_local_time?: string | null;
  latest_end_local_time_in_session?: string | null;
  latest_end_local_time_summer_break?: string | null;
  latest_end_local_time_school_day?: string | null;
  latest_end_local_time_non_school_day?: string | null;
  latest_end_local_time_preceding_non_school_day?: string | null;
};

export type EmployeeRoleAssignment = {
  id: string;
  employee_id: string;
  role_id: string;
  role_code?: string | null;
  role_name?: string | null;
  proficiency_level: number;
  is_primary: boolean;
  acquired_at?: string | null;
  role_metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type EmployeeLocationAssignment = {
  id: string;
  employee_id: string;
  location_id: string;
  location_name?: string | null;
  location_slug?: string | null;
  is_primary: boolean;
  access_level: string;
  location_source?: string | null;
  can_cover_last_minute: boolean;
  can_blast: boolean;
  travel_radius_miles?: number | null;
  location_metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type EmployeeProfile = EmployeeSummary & {
  roles: EmployeeRoleAssignment[];
  locations: EmployeeLocationAssignment[];
  work_permits: EmployeeWorkPermit[];
};

export type EmployeeWorkPermit = {
  id: string;
  employee_id: string;
  permit_number: string;
  issuing_authority?: string | null;
  issued_on?: string | null;
  effective_start_date?: string | null;
  effective_end_date?: string | null;
  max_daily_minutes?: number | null;
  max_weekly_minutes?: number | null;
  earliest_start_local_time?: string | null;
  latest_end_local_time?: string | null;
  rule_profile?: EmployeeWorkPermitRuleProfile | null;
  permit_metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type EmployeeWorkPermitTemplate = {
  code: string;
  label: string;
  description?: string | null;
  jurisdiction_code?: string | null;
  source_url?: string | null;
  rule_profile: EmployeeWorkPermitRuleProfile;
};

export type EmployeeDeleteReadiness = {
  business_id: string;
  employee_id: string;
  can_delete: boolean;
  reason?: string | null;
};

export type EmployeeAvailabilityRule = {
  id: string;
  employee_id: string;
  day_of_week: number;
  start_local_time: string;
  end_local_time: string;
  timezone: string;
  availability_type: string;
  valid_from?: string | null;
  valid_until?: string | null;
  priority: number;
  availability_metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type SelfEmployeeAvailability = {
  employee_id: string;
  employee_name: string;
  timezone: string;
  rules: EmployeeAvailabilityRule[];
};

export type EmployeeAvailabilityRulePayload = {
  day_of_week: number;
  start_local_time: string;
  end_local_time: string;
  timezone: string;
  availability_type?: string;
  valid_from?: string | null;
  valid_until?: string | null;
  priority?: number;
  availability_metadata?: Record<string, unknown>;
};

export type EmployeeRoleUpsertPayload = {
  role_id: string;
  proficiency_level?: number;
  is_primary?: boolean;
  role_metadata?: Record<string, unknown>;
};

export type EmployeeLocationUpsertPayload = {
  location_id: string;
  is_primary?: boolean;
  access_level?: string;
  location_source?: string | null;
  can_cover_last_minute?: boolean;
  can_blast?: boolean;
  travel_radius_miles?: number | null;
  location_metadata?: Record<string, unknown>;
};

export type EmployeeUpdatePayload = {
  full_name?: string | null;
  preferred_name?: string | null;
  phone_e164?: string | null;
  email?: string | null;
  external_ref?: string | null;
  employee_number?: string | null;
  base_hourly_rate_cents?: number | null;
  date_of_birth?: string | null;
  minor_school_status?: string | null;
  work_permit_number?: string | null;
  work_permit_effective_start_on?: string | null;
  work_permit_expires_on?: string | null;
  work_permit_max_daily_minutes?: number | null;
  work_permit_max_weekly_minutes?: number | null;
  work_permit_earliest_start_local_time?: string | null;
  work_permit_latest_end_local_time?: string | null;
  work_permits?: EmployeeWorkPermitPayload[];
  employment_type?: string | null;
  status?: string | null;
  hire_date?: string | null;
  termination_date?: string | null;
  notes?: string | null;
  employee_metadata?: Record<string, unknown>;
  notification_preferences?: Partial<EmployeeNotificationPreferences>;
  roles?: EmployeeRoleUpsertPayload[];
  locations?: EmployeeLocationUpsertPayload[];
};

export type EmployeeCreatePayload = {
  full_name: string;
  preferred_name?: string | null;
  phone_e164?: string | null;
  email?: string | null;
  external_ref?: string | null;
  employee_number?: string | null;
  base_hourly_rate_cents?: number | null;
  date_of_birth?: string | null;
  minor_school_status?: string | null;
  work_permit_number?: string | null;
  work_permit_effective_start_on?: string | null;
  work_permit_expires_on?: string | null;
  work_permit_max_daily_minutes?: number | null;
  work_permit_max_weekly_minutes?: number | null;
  work_permit_earliest_start_local_time?: string | null;
  work_permit_latest_end_local_time?: string | null;
  work_permits?: EmployeeWorkPermitPayload[];
  employment_type?: string | null;
  primary_location_id?: string | null;
  hire_date?: string | null;
  notes?: string | null;
  employee_metadata?: Record<string, unknown>;
};

export type EmployeeWorkPermitPayload = {
  permit_number: string;
  issuing_authority?: string | null;
  issued_on?: string | null;
  effective_start_date?: string | null;
  effective_end_date?: string | null;
  max_daily_minutes?: number | null;
  max_weekly_minutes?: number | null;
  earliest_start_local_time?: string | null;
  latest_end_local_time?: string | null;
  rule_profile?: EmployeeWorkPermitRuleProfile | null;
  permit_metadata?: Record<string, unknown>;
};

export type EmployeeImportError = {
  row_number?: number | null;
  message: string;
};

export type EmployeeBulkImportResponse = {
  created_count: number;
  skipped_count: number;
  employees: EmployeeSummary[];
  errors: EmployeeImportError[];
  default_location_id?: string | null;
  default_location_name?: string | null;
};

export async function listEmployees(
  businessId: string,
): Promise<EmployeeSummary[]> {
  const response = await apiFetchApp(`${API_PREFIX}/businesses/${businessId}/employees`);
  if (!response.ok) {
    throw new Error(await parseError(response));
  }
  return (await response.json()) as EmployeeSummary[];
}

export async function createEmployee(
  businessId: string,
  payload: EmployeeCreatePayload,
): Promise<EmployeeSummary> {
  const response = await apiFetchApp(`${API_PREFIX}/businesses/${businessId}/employees`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!response.ok) {
    throw new Error(await parseError(response));
  }
  return (await response.json()) as EmployeeSummary;
}

export async function getEmployeeProfile(
  businessId: string,
  employeeId: string,
): Promise<EmployeeProfile> {
  const response = await apiFetchApp(
    `${API_PREFIX}/businesses/${businessId}/employees/${employeeId}`,
  );
  if (!response.ok) {
    throw new Error(await parseError(response));
  }
  return (await response.json()) as EmployeeProfile;
}

export async function listWorkPermitTemplates(
  businessId: string,
): Promise<EmployeeWorkPermitTemplate[]> {
  const response = await apiFetchApp(
    `${API_PREFIX}/businesses/${businessId}/employees/work-permit-templates`,
  );
  if (!response.ok) {
    throw new Error(await parseError(response));
  }
  return (await response.json()) as EmployeeWorkPermitTemplate[];
}

export async function updateEmployee(
  businessId: string,
  employeeId: string,
  payload: EmployeeUpdatePayload,
): Promise<EmployeeProfile> {
  const response = await apiFetchApp(
    `${API_PREFIX}/businesses/${businessId}/employees/${employeeId}`,
    {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    },
  );
  if (!response.ok) {
    throw new Error(await parseError(response));
  }
  return (await response.json()) as EmployeeProfile;
}

export async function getEmployeeDeleteReadiness(
  businessId: string,
  employeeId: string,
): Promise<EmployeeDeleteReadiness> {
  const response = await apiFetchApp(
    `${API_PREFIX}/businesses/${businessId}/employees/${employeeId}/delete-readiness`,
  );
  if (!response.ok) {
    throw new Error(await parseError(response));
  }
  return (await response.json()) as EmployeeDeleteReadiness;
}

export async function deleteEmployee(
  businessId: string,
  employeeId: string,
): Promise<{ deleted: boolean; employee_id: string }> {
  const response = await apiFetchApp(
    `${API_PREFIX}/businesses/${businessId}/employees/${employeeId}`,
    {
      method: "DELETE",
    },
  );
  if (!response.ok) {
    throw new Error(await parseError(response));
  }
  return (await response.json()) as { deleted: boolean; employee_id: string };
}

export async function getSelfEmployeeAvailability(
  businessId: string,
): Promise<SelfEmployeeAvailability> {
  const response = await apiFetchApp(
    `${API_PREFIX}/businesses/${businessId}/employees/availability-rules/self`,
  );
  if (!response.ok) {
    throw new Error(await parseError(response));
  }
  return (await response.json()) as SelfEmployeeAvailability;
}

export async function getEmployeeAvailability(
  businessId: string,
  employeeId: string,
): Promise<SelfEmployeeAvailability> {
  const response = await apiFetchApp(
    `${API_PREFIX}/businesses/${businessId}/employees/${employeeId}/availability-rules`,
  );
  if (!response.ok) {
    throw new Error(await parseError(response));
  }
  return (await response.json()) as SelfEmployeeAvailability;
}

export async function replaceSelfEmployeeAvailability(
  businessId: string,
  payload: { rules: EmployeeAvailabilityRulePayload[] },
): Promise<SelfEmployeeAvailability> {
  const response = await apiFetchApp(
    `${API_PREFIX}/businesses/${businessId}/employees/availability-rules/self`,
    {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    },
  );
  if (!response.ok) {
    throw new Error(await parseError(response));
  }
  return (await response.json()) as SelfEmployeeAvailability;
}

export async function replaceEmployeeAvailability(
  businessId: string,
  employeeId: string,
  payload: { rules: EmployeeAvailabilityRulePayload[] },
): Promise<SelfEmployeeAvailability> {
  const response = await apiFetchApp(
    `${API_PREFIX}/businesses/${businessId}/employees/${employeeId}/availability-rules`,
    {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    },
  );
  if (!response.ok) {
    throw new Error(await parseError(response));
  }
  return (await response.json()) as SelfEmployeeAvailability;
}

export async function importEmployees(
  businessId: string,
  file: File,
): Promise<EmployeeBulkImportResponse> {
  const formData = new FormData();
  formData.append("file", file);

  const response = await apiFetchApp(
    `${API_PREFIX}/businesses/${businessId}/employees/import`,
    {
      method: "POST",
      body: formData,
    },
  );
  if (!response.ok) {
    throw new Error(await parseError(response));
  }
  return (await response.json()) as EmployeeBulkImportResponse;
}

export async function downloadEmployeeImportTemplate(
  businessId: string,
): Promise<void> {
  const response = await apiFetchApp(
    `${API_PREFIX}/businesses/${businessId}/employees/import/template`,
  );
  if (!response.ok) {
    throw new Error(await parseError(response));
  }

  const blob = await response.blob();
  const url = URL.createObjectURL(blob);
  const link = document.createElement("a");
  link.href = url;
  link.download = "backfill-employee-roster-template.csv";
  document.body.appendChild(link);
  link.click();
  link.remove();
  URL.revokeObjectURL(url);
}

File: app/schemas/workforce.py
from __future__ import annotations

from datetime import date, datetime, time
from typing import Literal, Optional
from uuid import UUID

from pydantic import Field

from app.models.common import EmployeeStatus
from app.schemas.common import BaseSchema

WeekdayName = Literal[
    "sunday",
    "monday",
    "tuesday",
    "wednesday",
    "thursday",
    "friday",
    "saturday",
]


class EmployeeWorkPermitRuleProfile(BaseSchema):
    template_code: Optional[str] = None
    jurisdiction_code: Optional[str] = None
    source_url: Optional[str] = None
    allowed_weekdays: list[WeekdayName] = Field(default_factory=list)
    daily_max_minutes: Optional[int] = Field(default=None, ge=0)
    daily_max_minutes_in_session: Optional[int] = Field(default=None, ge=0)
    daily_max_minutes_summer_break: Optional[int] = Field(default=None, ge=0)
    daily_max_minutes_school_day: Optional[int] = Field(default=None, ge=0)
    daily_max_minutes_non_school_day: Optional[int] = Field(default=None, ge=0)
    daily_max_minutes_preceding_non_school_day: Optional[int] = Field(default=None, ge=0)
    weekly_max_minutes: Optional[int] = Field(default=None, ge=0)
    weekly_max_minutes_in_session: Optional[int] = Field(default=None, ge=0)
    weekly_max_minutes_summer_break: Optional[int] = Field(default=None, ge=0)
    weekly_max_minutes_school_week: Optional[int] = Field(default=None, ge=0)
    weekly_max_minutes_non_school_week: Optional[int] = Field(default=None, ge=0)
    earliest_start_local_time: Optional[time] = None
    earliest_start_local_time_school_day: Optional[time] = None
    earliest_start_local_time_non_school_day: Optional[time] = None
    latest_end_local_time: Optional[time] = None
    latest_end_local_time_in_session: Optional[time] = None
    latest_end_local_time_summer_break: Optional[time] = None
    latest_end_local_time_school_day: Optional[time] = None
    latest_end_local_time_non_school_day: Optional[time] = None
    latest_end_local_time_preceding_non_school_day: Optional[time] = None


class EmployeeWorkPermitCreate(BaseSchema):
    permit_number: str
    issuing_authority: Optional[str] = None
    issued_on: Optional[date] = None
    effective_start_date: Optional[date] = None
    effective_end_date: Optional[date] = None
    max_daily_minutes: Optional[int] = Field(default=None, ge=0)
    max_weekly_minutes: Optional[int] = Field(default=None, ge=0)
    earliest_start_local_time: Optional[time] = None
    latest_end_local_time: Optional[time] = None
    rule_profile: Optional[EmployeeWorkPermitRuleProfile] = None
    permit_metadata: dict = Field(default_factory=dict)


class EmployeeWorkPermitRead(BaseSchema):
    id: UUID
    employee_id: UUID
    permit_number: str
    issuing_authority: Optional[str] = None
    issued_on: Optional[date] = None
    effective_start_date: Optional[date] = None
    effective_end_date: Optional[date] = None
    max_daily_minutes: Optional[int] = None
    max_weekly_minutes: Optional[int] = None
    earliest_start_local_time: Optional[time] = None
    latest_end_local_time: Optional[time] = None
    rule_profile: Optional[EmployeeWorkPermitRuleProfile] = None
    permit_metadata: dict
    created_at: datetime
    updated_at: datetime


class EmployeeWorkPermitTemplateRead(BaseSchema):
    code: str
    label: str
    description: Optional[str] = None
    jurisdiction_code: Optional[str] = None
    source_url: Optional[str] = None
    source_document_title: Optional[str] = None
    source_version: Optional[str] = None
    source_hash: Optional[str] = None
    effective_start_date: Optional[date] = None
    effective_end_date: Optional[date] = None
    payload_hash: Optional[str] = None
    rule_families: list[str] = Field(default_factory=list)
    rule_profile: EmployeeWorkPermitRuleProfile


class EmployeeCreate(BaseSchema):
    full_name: str
    preferred_name: Optional[str] = None
    phone_e164: Optional[str] = None
    email: Optional[str] = None
    external_ref: Optional[str] = None
    employee_number: Optional[str] = None
    employment_type: Optional[str] = None
    base_hourly_rate_cents: Optional[int] = Field(default=None, ge=0)
    date_of_birth: Optional[date] = None
    minor_school_status: Optional[str] = None
    work_permit_number: Optional[str] = None
    work_permit_effective_start_on: Optional[date] = None
    work_permit_expires_on: Optional[date] = None
    work_permit_max_daily_minutes: Optional[int] = Field(default=None, ge=0)
    work_permit_max_weekly_minutes: Optional[int] = Field(default=None, ge=0)
    work_permit_earliest_start_local_time: Optional[time] = None
    work_permit_latest_end_local_time: Optional[time] = None
    work_permits: list[EmployeeWorkPermitCreate] = Field(default_factory=list)
    primary_location_id: Optional[UUID] = None
    hire_date: Optional[date] = None
    notes: Optional[str] = None
    employee_metadata: dict = Field(default_factory=dict)


class EmployeeNotificationPreferencesRead(BaseSchema):
    schedule_publish_email_enabled: bool = True
    schedule_publish_sms_enabled: bool = False
    email_opted_out_at: Optional[datetime] = None
    sms_opted_out_at: Optional[datetime] = None
    email_opt_out_reason: Optional[str] = None
    sms_opt_out_reason: Optional[str] = None


class EmployeeNotificationPreferencesUpdate(BaseSchema):
    schedule_publish_email_enabled: Optional[bool] = None
    schedule_publish_sms_enabled: Optional[bool] = None
    email_opted_out_at: Optional[datetime] = None
    sms_opted_out_at: Optional[datetime] = None
    email_opt_out_reason: Optional[str] = None
    sms_opt_out_reason: Optional[str] = None


class EmployeeRead(BaseSchema):
    id: UUID
    business_id: UUID
    primary_location_id: Optional[UUID]
    primary_location_name: Optional[str] = None
    primary_role_id: Optional[UUID] = None
    primary_role_name: Optional[str] = None
    external_ref: Optional[str]
    employee_number: Optional[str]
    full_name: str
    preferred_name: Optional[str]
    phone_e164: Optional[str]
    email: Optional[str]
    base_hourly_rate_cents: Optional[int] = None
    date_of_birth: Optional[date] = None
    minor_school_status: Optional[str] = None
    work_permit_number: Optional[str] = None
    work_permit_effective_start_on: Optional[date] = None
    work_permit_expires_on: Optional[date] = None
    work_permit_max_daily_minutes: Optional[int] = None
    work_permit_max_weekly_minutes: Optional[int] = None
    work_permit_earliest_start_local_time: Optional[time] = None
    work_permit_latest_end_local_time: Optional[time] = None
    reliability_score: Optional[float] = 0.7
    status: str
    employment_type: Optional[str]
    hire_date: Optional[date]
    termination_date: Optional[date]
    notes: Optional[str]
    employee_metadata: dict
    notification_preferences: EmployeeNotificationPreferencesRead = Field(
        default_factory=EmployeeNotificationPreferencesRead
    )
    role_ids: list[UUID] = Field(default_factory=list)
    role_names: list[str] = Field(default_factory=list)
    location_ids: list[UUID] = Field(default_factory=list)
    location_names: list[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class EmployeeEnrollAtLocationCreate(BaseSchema):
    location_id: UUID
    role_ids: list[UUID]
    full_name: str
    preferred_name: Optional[str] = None
    phone_e164: Optional[str] = None
    email: Optional[str] = None
    external_ref: Optional[str] = None
    employee_number: Optional[str] = None
    employment_type: Optional[str] = None
    base_hourly_rate_cents: Optional[int] = Field(default=None, ge=0)
    date_of_birth: Optional[date] = None
    minor_school_status: Optional[str] = None
    work_permit_number: Optional[str] = None
    work_permit_effective_start_on: Optional[date] = None
    work_permit_expires_on: Optional[date] = None
    work_permit_max_daily_minutes: Optional[int] = Field(default=None, ge=0)
    work_permit_max_weekly_minutes: Optional[int] = Field(default=None, ge=0)
    work_permit_earliest_start_local_time: Optional[time] = None
    work_permit_latest_end_local_time: Optional[time] = None
    work_permits: list[EmployeeWorkPermitCreate] = Field(default_factory=list)
    hire_date: Optional[date] = None
    notes: Optional[str] = None
    employee_metadata: dict = Field(default_factory=dict)


class EmployeeEnrollmentRead(BaseSchema):
    employee: EmployeeRead
    roles: list["EmployeeRoleRead"]


class EmployeeImportErrorRead(BaseSchema):
    row_number: Optional[int] = None
    message: str


class EmployeeBulkImportRead(BaseSchema):
    created_count: int
    skipped_count: int
    employees: list[EmployeeRead] = Field(default_factory=list)
    errors: list[EmployeeImportErrorRead] = Field(default_factory=list)
    default_location_id: Optional[UUID] = None
    default_location_name: Optional[str] = None


class EmployeeRoleCreate(BaseSchema):
    role_id: UUID
    proficiency_level: int = 1
    is_primary: bool = False
    role_metadata: dict = Field(default_factory=dict)


class EmployeeRoleRead(BaseSchema):
    id: UUID
    employee_id: UUID
    role_id: UUID
    role_code: Optional[str] = None
    role_name: Optional[str] = None
    proficiency_level: int
    is_primary: bool
    acquired_at: Optional[datetime]
    role_metadata: dict
    created_at: datetime
    updated_at: datetime


class EmployeeLocationCreate(BaseSchema):
    location_id: UUID
    is_primary: bool = False
    access_level: str = "approved"
    location_source: Optional[str] = None
    can_cover_last_minute: bool = True
    can_blast: bool = True
    travel_radius_miles: Optional[int] = None
    location_metadata: dict = Field(default_factory=dict)


class EmployeeLocationRead(BaseSchema):
    id: UUID
    employee_id: UUID
    location_id: UUID
    location_name: Optional[str] = None
    location_slug: Optional[str] = None
    is_primary: bool
    access_level: str
    location_source: Optional[str]
    can_cover_last_minute: bool
    can_blast: bool
    travel_radius_miles: Optional[int]
    location_metadata: dict
    created_at: datetime
    updated_at: datetime


class EmployeeRoleUpsert(BaseSchema):
    role_id: UUID
    proficiency_level: Optional[int] = None
    is_primary: bool = False
    role_metadata: Optional[dict] = None


class EmployeeLocationUpsert(BaseSchema):
    location_id: UUID
    is_primary: bool = False
    access_level: Optional[str] = None
    location_source: Optional[str] = None
    can_cover_last_minute: Optional[bool] = None
    can_blast: Optional[bool] = None
    travel_radius_miles: Optional[int] = None
    location_metadata: Optional[dict] = None


class EmployeeUpdate(BaseSchema):
    full_name: Optional[str] = None
    preferred_name: Optional[str] = None
    phone_e164: Optional[str] = None
    email: Optional[str] = None
    external_ref: Optional[str] = None
    employee_number: Optional[str] = None
    employment_type: Optional[str] = None
    base_hourly_rate_cents: Optional[int] = Field(default=None, ge=0)
    date_of_birth: Optional[date] = None
    minor_school_status: Optional[str] = None
    work_permit_number: Optional[str] = None
    work_permit_effective_start_on: Optional[date] = None
    work_permit_expires_on: Optional[date] = None
    work_permit_max_daily_minutes: Optional[int] = Field(default=None, ge=0)
    work_permit_max_weekly_minutes: Optional[int] = Field(default=None, ge=0)
    work_permit_earliest_start_local_time: Optional[time] = None
    work_permit_latest_end_local_time: Optional[time] = None
    work_permits: Optional[list[EmployeeWorkPermitCreate]] = None
    status: Optional[EmployeeStatus] = None
    hire_date: Optional[date] = None
    termination_date: Optional[date] = None
    notes: Optional[str] = None
    employee_metadata: Optional[dict] = None
    notification_preferences: Optional[EmployeeNotificationPreferencesUpdate] = None
    roles: Optional[list[EmployeeRoleUpsert]] = None
    locations: Optional[list[EmployeeLocationUpsert]] = None


class EmployeeProfileRead(EmployeeRead):
    roles: list[EmployeeRoleRead] = Field(default_factory=list)
    locations: list[EmployeeLocationRead] = Field(default_factory=list)
    work_permits: list[EmployeeWorkPermitRead] = Field(default_factory=list)


class EmployeeDeleteReadinessRead(BaseSchema):
    business_id: UUID
    employee_id: UUID
    can_delete: bool
    reason: Optional[str] = None


class EmployeeDeleteResponse(BaseSchema):
    deleted: bool
    employee_id: UUID


class EmployeeAvailabilityRuleCreate(BaseSchema):
    day_of_week: int
    start_local_time: time
    end_local_time: time
    timezone: str
    availability_type: str = "available"
    valid_from: Optional[date] = None
    valid_until: Optional[date] = None
    priority: int = 0
    availability_metadata: dict = Field(default_factory=dict)


class EmployeeAvailabilityRuleRead(BaseSchema):
    id: UUID
    employee_id: UUID
    day_of_week: int
    start_local_time: time
    end_local_time: time
    timezone: str
    availability_type: str
    valid_from: Optional[date]
    valid_until: Optional[date]
    priority: int
    availability_metadata: dict
    created_at: datetime
    updated_at: datetime


class EmployeeAvailabilityRuleReplace(BaseSchema):
    rules: list[EmployeeAvailabilityRuleCreate] = Field(default_factory=list)


class SelfEmployeeAvailabilityRead(BaseSchema):
    employee_id: UUID
    employee_name: str
    timezone: str
    rules: list[EmployeeAvailabilityRuleRead] = Field(default_factory=list)
